common/fs/src/cache/mod.rs

```diff
@@ -1,47 +1,56 @@
-use crate::cache::entry::{Entry, EntryPtr};
+use crate::cache::entry::Entry;
 use crate::cache::event::Event;
 use crate::cache::watch::{WatchEvent, Watcher};
 use crate::rule::{GlobRule, Rules, Status};
-use futures::{Stream, StreamExt};
-use hashbrown::hash_map::Entry as HashMapEntry;
-use hashbrown::HashMap;
-use inotify::WatchDescriptor;
-use metrics::Metrics;
+
+use std::cell::RefCell;
 use std::ffi::OsString;
 use std::fmt;
 use std::fs::read_dir;
 use std::fs::OpenOptions;
 use std::iter::FromIterator;
-use std::ops::{Deref, DerefMut};
+use std::ops::Deref;
 use std::path::{Component, PathBuf};
-use std::ptr::NonNull;
+use std::rc::Rc;
 use std::sync::{Arc, Mutex};
 
+use futures::{Stream, StreamExt};
+use hashbrown::hash_map::Entry as HashMapEntry;
+use hashbrown::HashMap;
+use inotify::WatchDescriptor;
+use metrics::Metrics;
+use slotmap::{DefaultKey, SlotMap};
+
 pub mod dir_path;
 pub mod entry;
 pub mod event;
 pub use dir_path::{DirPathBuf, DirPathBufError};
 
 mod watch;
 
-type Children<T> = HashMap<OsString, Box<Entry<T>>>;
-type Symlinks<T> = HashMap<PathBuf, Vec<EntryPtr<T>>>;
-type WatchDescriptors<T> = HashMap<WatchDescriptor, Vec<EntryPtr<T>>>;
+type Children = HashMap<OsString, EntryKey>;
+type Symlinks = HashMap<PathBuf, Vec<EntryKey>>;
+type WatchDescriptors = HashMap<WatchDescriptor, Vec<EntryKey>>;
+
+pub type EntryKey = DefaultKey;
+
+type EntryMap<T> = SlotMap<EntryKey, RefCell<entry::Entry<T>>>;
 
 pub struct FileSystem<T>
 where
     T: Clone + std::fmt::Debug,
 {
     watcher: Watcher,
-    root: Box<Entry<T>>,
+    pub entries: Rc<RefCell<EntryMap<T>>>,
```

Comment:
The other motivation for having this here was to avoid the need to pass &mut entries into the process* methods, which would have meant leaking the fact the storage is out of line right up into the tailer.

This way stow the reference in the FileSystem type and cheaply grab a reference each time process is called